mava/components/jax/training/trainer.py

```diff
@@ -113,9 +145,197 @@ def on_training_utility_fns(self, trainer: SystemTrainer) -> None:
+        # Setup specs for each network
+        builder.store.net_spec_keys = {}
+        for i in range(len(unique_net_keys)):
+            builder.store.net_spec_keys[unique_net_keys[i]] = builder.store.agents[
+                i % len(builder.store.agents)
+            ]
+
+        # Setup table_network_config
+        builder.store.table_network_config = {}
+        for trainer_key in builder.store.trainer_networks.keys():
+            most_matches = 0
+            trainer_nets = builder.store.trainer_networks[trainer_key]
+            for sample in builder.store.network_sampling_setup:
+                matches = 0
+                for entry in sample:
+                    if entry in trainer_nets:
+                        matches += 1
+                if most_matches < matches:
+                    matches = most_matches
+                    builder.store.table_network_config[trainer_key] = sample
+
+        builder.store.networks = builder.store.network_factory()
+
+    def on_training_utility_fns(self, trainer: SystemTrainer) -> None:
         """_summary_"""
-        return "trainer_init"
+        # Convert network keys for the trainer.
+        trainer.store.trainer_table_entry = trainer.store.table_network_config[
+            trainer.store.trainer_id
+        ]
+        trainer.store.trainer_agents = trainer.store.agents[
+            : len(trainer.store.trainer_table_entry)
+        ]
+        trainer.store.trainer_agent_net_keys = {
+            agent: trainer.store.trainer_table_entry[a_i]
+            for a_i, agent in enumerate(trainer.store.trainer_agents)
+        }
+
+    @staticmethod
+    def config_class() -> Optional[Callable]:
+        """Config class used for component.
+
+        Returns:
+            config class/dataclass for component.
+        """
+        return OneTrainerPerNetworkInitConfig
+
+
+@dataclass
+class CustomTrainerInitConfig(BaseTrainerInitConfig):
+    pass
+
+
+class CustomTrainerInit(BaseTrainerInit):
+    def __init__(self, config: CustomTrainerInitConfig = CustomTrainerInitConfig()):
+        """Initialises custom trainers.
+
+        Here a custom trainer network configuration can be given as a dictionary
+        assigning specific trainers to specific networks.
+
+        Args:
+            config : a dataclass specifying the component parameters.
+        """
+
+        self.config = config
+
+    def on_building_init_end(self, builder: SystemBuilder) -> None:
+        """Assigns trainers to networks for training.
+
+        Args:
+            builder : the system builder
+        Raises:
+            ValueError: Raises an error when trainer_networks is not
+                        passed in as a dictionary.
+        """
+        trainer_networks = self.config.trainer_networks
+        unique_net_keys = builder.store.unique_net_keys
+
+        # Setup trainer_networks
+        if not isinstance(trainer_networks, dict) or trainer_networks == {}:
```

Comment:
More of a question but why do we use dict in (various places in) mava instead of just using {} notation (from what I read online, dict is slower)